references/classification/train.py

```diff
@@ -170,23 +170,23 @@ def main(args):
     if args.distributed and args.sync_bn:
         model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
 
-    model_without_ddp = model
-    if args.distributed:
-        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
-        model_without_ddp = model.module
-
     criterion = nn.CrossEntropyLoss()
 
     optimizer = torch.optim.SGD(
         model.parameters(), lr=args.lr, momentum=args.momentum, weight_decay=args.weight_decay)
 
-    lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=args.lr_step_size, gamma=args.lr_gamma)
-
     if args.apex:
         model, optimizer = amp.initialize(model, optimizer,
                                           opt_level=args.apex_opt_level
                                           )
 
+    lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=args.lr_step_size, gamma=args.lr_gamma)
+
+    model_without_ddp = model
+    if args.distributed:
```

Comment:
I'm not very comfortable moving this block after having creating the `optimizer`.

The reason is that I'm not sure if there are any guarantees that the references to the original parameters of the model will be the same before and after the DDP has been applied.
This means that, potentially, the optimizer would be seeing an old set of parameters from the model, and training would not work as expected.
A classical example is when the user moves the model to the GPU after having constructing the optimizer, which still points to the CPU tensors, and training doesn't happen at all.

@pietern do you think this concern is justified or do we have guarantees that DDP (in all its flavours, single or multi GPU per process) will not modify the references to the original parameters after construction?

Reply:
I think it's a valid concern. DDP doesn't modify the model it wraps. It only replicates it in the case of using multiple replicas per process. But it never mutates the underlying model.

Regardless, you'll have a problem when you move it to GPU, or change precision, or do any other destructive mutation. To compartmentalize, I think it's best to just delay creating the optimizer until you've done all of this, including wrapping in DDP.

Reply:
According to https://github.com/NVIDIA/apex/tree/master/examples/imagenet
```
To use DDP with apex.amp, the only gotcha is that

model, optimizer = amp.initialize(model, optimizer, flags...)

must precede

model = DDP(model)
```
 
I have tested this ordering, i.e. DDP **after** optimizer creation & APEX wrapping, till (almost) convergence with 4 GPUs:

`Epoch: [236]  [1000/1252]  eta: 0:01:31  lr: 0.0003902476545030211  img/s: 788.6827109884629  loss: 0.5897 (0.6214)  acc1: 85.5469 (85.1801)  acc5: 94.9219 (94.7428)  time: 0.3561  data: 0.0002  max mem: 14467
`

Reply:
Next I'm testing DDP **after** optimizer creation but **without** APEX on 4 GPUs. A few initial iterations shows that model is learning well.

Reply:
Please let us know how it goes

Reply:
mid way through 4-GPU training, seems like it converging just fine
`Epoch: [39]  [1300/2503]  eta: 0:07:41  lr: 0.020883504974794645  img/s: 345.68365366660873  loss: 1.3453 (1.3118)  acc1: 67.1875 (68.6227)  acc5: 85.9375 (87.3265)  time: 0.3707  data: 0.0002  max mem: 13502`

Reply:
my training was cut short at epoch 80 due to out of disk quota error.
`Epoch: [80]  [2500/2503]  eta: 0:00:01  lr: 0.009121630871114108  img/s: 346.88059463297736  loss: 1.0024 (0.9745)  acc1: 75.0000 (76.1389)  acc5: 90.6250 (91.3506)  time: 0.8924  data: 0.0001  max mem: 13502
`
but overall empirical evidence suggests that moving  DDP  to after optimizer creation works just fine (with or without APEX).